src/Compilers/CSharp/Portable/Binder/Lifetime.cs

```diff
@@ -0,0 +1,130 @@
+        ///</summary>
+        public static Lifetime Empty => new Lifetime(uint.MaxValue);
+
+        /// <summary>
+        /// Gets a lifetime which is narrower than the given lifetime.
+        /// Used to "enter" a nested local scope.
+        /// </summary>
+        public Lifetime Narrower()
+        {
+            var result = new Lifetime(this._value + 1);
+            // Narrower() operator should always result in a local lifetime
+            Debug.Assert(!result.IsReturnable);
+            return result;
+        }
+
+        /// <summary>
+        /// Gets a lifetime which is wider than the given lifetime.
+        /// Used to "exit" a nested local scope.
+        /// </summary>
+        public Lifetime Wider()
+        {
+            // Wider() operator should always start from a local lifetime
+            Debug.Assert(!this.IsReturnable);
+            return new Lifetime(this._value - 1);
+        }
+
+        public bool IsCallingMethod => _value == CallingMethodRaw;
+        public bool IsReturnOnly => _value == ReturnOnlyRaw;
+        public bool IsReturnable => _value is CallingMethodRaw or ReturnOnlyRaw;
+
+        /// <summary>Returns true if a 'ref' with this lifetime can be converted to the 'other' lifetime. Otherwise, returns false.</summary>
+        /// <remarks>Generally, a wider lifetime is convertible to a narrower lifetime.</remarks>
+        public bool IsConvertibleTo(Lifetime other)
+            => this._value <= other._value;
+
+        /// <summary>
+        /// Returns the narrower of two lifetimes.
+        /// </summary>
+        /// <remarks>
+        /// In other words, this method returns the widest lifetime which 'this' and 'other' are both convertible to.
+        /// If in future we added the concept of unrelated lifetimes (e.g. to implement 'ref scoped'), this method would perhaps return a Nullable,
+        /// for the case that no lifetime exists which both input lifetimes are convertible to.
+        /// </remarks>
+        public Lifetime Intersect(Lifetime other)
+            => this.IsConvertibleTo(other) ? other : this;
+
+        /// <summary>
+        /// Returns the wider of two lifetimes.
+        /// </summary>
+        /// <remarks>In other words, this method returns the narrowest lifetime which can be converted to both 'this' and 'other'.</remarks>
+        public Lifetime Union(Lifetime other)
+            => this.IsConvertibleTo(other) ? this : other;
+
+        /// <summary>Returns true if this lifetime is the same as 'other' (i.e. for invariant nested conversion).</summary>
+        public bool Equals(Lifetime other)
+            => this._value == other._value;
+
+        public override bool Equals(object? obj)
+            => obj is Lifetime other && this.Equals(other);
+
+        public override int GetHashCode()
+            => unchecked((int)_value);
+
+        public static bool operator ==(Lifetime lhs, Lifetime rhs)
```

Comment:
I found that BoundNodes generation expects this operator to exist for value types used as fields. So I ended up filling in all the equality methods rather than just having an 'Equals(Lifetime)'.

I did avoid using this operator in manually written code but don't feel extremely strongly about that.